### PR TITLE
add gibberish & repetition filtering

### PR DESCRIPTION
**Pull request by hallerite**
This PR adds a new `RolloutFilter` class that allows one to define rules for filtering out rollouts, for example when they contain gibberish or repetitions.

---

> [!NOTE]
> **Medium Risk**
> Touches the training data path by mutating rollout masks/stop conditions before advantage computation, which can change what samples reach the trainer. Logic is localized and covered by unit tests, but misconfiguration could silently drop learning signal.
> 
> **Overview**
> Adds **orchestrator-side rollout filtering** to detect degenerate generations and optionally prevent them from contributing to training.
> 
> Introduces configurable `orchestrator.filters` with `gibberish` (rare tokens at high entropy) and `repetition` (long high-confidence streaks) detectors, wires them into the orchestrator loop to zero `completion_mask`/set `stop_condition` when `enforce=True`, and logs aggregate filter metrics. Also removes the previous Chinese-character counting metrics/utilities and adds unit tests covering filter detection, enforcement behavior, and setup.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit b1dd7e796a384c14073d71e7aced51771956ffcc. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/prime_rl/orchestrator/filters.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+    for rollout in rollouts:
+        for filt in filters:
+            result = filt.check(rollout)
+            if result.detected:
+                rollout["reward"] = 0.0
+                for step in rollout["trajectory"]:
+                    tokens = step["tokens"]
+                    if tokens is not None:
+                        tokens["completion_mask"] = [0] * len(tokens["completion_mask"])
```
[the remark this answers is not in the document]
Comment (hallerite) [reply src/prime_rl/orchestrator/filters.py]: I believe this is fine, as this is also how we handle truncation.

File: src/prime_rl/orchestrator/config.py [hunk 2]
```diff
@@ -561,6 +561,54 @@ def _advantage_config_discriminator(v: Any) -> str:
...
+def _filter_config_discriminator(v: Any) -> str:
+    if isinstance(v, dict):
+        return v.get("type", "gibberish")
+    return getattr(v, "type", "gibberish")
+
+
+FilterConfigType: TypeAlias = Annotated[
+    Annotated[GibberishFilterConfig, Tag("gibberish")] | Annotated[RepetitionFilterConfig, Tag("repetition")],
+    Discriminator(_filter_config_discriminator),
+]
```
Comment (samsja) [inline src/prime_rl/orchestrator/config.py]: we don't need tag here, can just do normal distriminator and let the user always precise with type

File: src/prime_rl/orchestrator/config.py [hunk 3]
```diff
@@ -561,6 +561,54 @@ def _advantage_config_discriminator(v: Any) -> str:
...
+
+
+class RepetitionFilterConfig(BaseConfig):
+    """Flags pathological repetition loops (Section 3.2, https://arxiv.org/abs/2506.13585)."""
+
+    type: Literal["repetition"] = "repetition"
+    window: Annotated[
+        int,
+        Field(ge=1, description="Number of consecutive high-probability steps before flagging."),
+    ] = 3_000
```
Comment (samsja) [inline src/prime_rl/orchestrator/config.py:621]: 3000 seems to high as a default ?

File: src/prime_rl/orchestrator/filters.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+@dataclass
+class FilterResult:
+    detected: bool
+    detection_index: int | None = None
+
+
+class RolloutFilter(Protocol):
+    name: str
+
+    def check(self, rollout: vf.RolloutOutput) -> FilterResult: ...
```
Comment (samsja) [inline src/prime_rl/orchestrator/filters.py:30]: FINALY SOMEONE USING PROTOCOL

File: src/prime_rl/orchestrator/config.py (hunk 3 above)
Comment (faresobeid) [reply src/prime_rl/orchestrator/config.py:621]: It's what minimax m1 has as a default image

File: tests/unit/orchestrator/test_filters.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+    }
+
+
+# --- GibberishFilter tests ---
+
+
+def test_gibberish_detects_rare_low_prob_token():
+    vocab_size = 128_000
+    logprob_threshold = -math.log(vocab_size) - 2.0
+    f = GibberishFilter(name="gibberish", token_id_threshold=100_000, logprob_threshold=logprob_threshold)
```
Comment (samsja) [inline tests/unit/orchestrator/test_filters.py]: I don;t like variable name withj one letter even in tests, word are cheap in the ai era

**Comment by faresobeid** [comment]
Nice work! Maybe we can have an option to track always but have the filtering optional so we can judge whether we should turn it on or not if we didn't already.

File: src/prime_rl/orchestrator/filters.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        return {}
+
+    counts: dict[str, int] = {f.name: 0 for f in filters}
+    total_filtered = 0
+
+    for rollout in rollouts:
+        for filt in filters:
+            result = filt.check(rollout)
+            if result.detected:
+                rollout["reward"] = 0.0
```
Comment (samsja) [inline src/prime_rl/orchestrator/filters.py]: what if reward is between -1 and 1 ?

Comment (hallerite) [reply src/prime_rl/orchestrator/filters.py]: we already mask out the gradient, so I guess we could simply not override the reward at all to get an undistorted group mean

File: src/prime_rl/orchestrator/config.py [hunk 7]
```diff
@@ -561,6 +561,64 @@ def _advantage_config_discriminator(v: Any) -> str:
 ]
 
 
+class GibberishFilterConfig(BaseModel):
+    """Flags rare tokens generated at high entropy (Section 5.2, https://arxiv.org/abs/2510.02387)."""
+
+    model_config = ConfigDict(extra="forbid")
```
Comment (mikasenghaas) [inline src/prime_rl/orchestrator/config.py:585]: this bc we cannot use base settings on discriminated unions right?

File: src/prime_rl/orchestrator/config.py [hunk 8]
```diff
@@ -561,6 +561,64 @@ def _advantage_config_discriminator(v: Any) -> str:
...
+
+    type: Literal["gibberish"] = "gibberish"
+    enforce: Annotated[
+        bool,
+        Field(
+            description="If True, zero reward and mask for detected rollouts. If False, only track detection metrics."
```
Comment (mikasenghaas) [inline src/prime_rl/orchestrator/config.py]: zero reward and mask? if we mask reward doesn't matter right?

File: src/prime_rl/orchestrator/config.py [hunk 9]
```diff
@@ -561,6 +561,64 @@ def _advantage_config_discriminator(v: Any) -> str:
...
+        ),
+    ] = False
+    token_id_threshold: Annotated[
```
Comment (mikasenghaas) [inline src/prime_rl/orchestrator/config.py:594]: not sure how much modern vocab sizes but could it make sense to make this a percentile? not too strong opinons here tho

File: src/prime_rl/orchestrator/config.py [hunk 10]
```diff
@@ -561,6 +561,64 @@ def _advantage_config_discriminator(v: Any) -> str:
...
+        int,
+        Field(description="Token IDs above this are candidates for gibberish. BPE tokens are sorted by merge order."),
+    ] = 100_000
+    logprob_offset: Annotated[
+        float,
+        Field(description="Offset from uniform distribution logprob. Threshold = -log(vocab_size) - logprob_offset."),
+    ] = 2.0
+    vocab_size: Annotated[
```
Comment (mikasenghaas) [inline src/prime_rl/orchestrator/config.py]: why would i ever change this value?

File: src/prime_rl/orchestrator/config.py [hunk 11]
```diff
@@ -561,6 +561,64 @@ def _advantage_config_discriminator(v: Any) -> str:
...
+    window: Annotated[
+        int,
+        Field(ge=1, description="Number of consecutive high-probability steps before flagging."),
+    ] = 3_000
+    prob_threshold: Annotated[
+        float,
+        Field(
+            gt=0,
+            le=1,
+            description="Token probability threshold. Steps where sampled prob exceeds this count toward the window.",
```
Comment (mikasenghaas) [inline src/prime_rl/orchestrator/config.py]: im not sure i get this description, can we make it clearer?

File: src/prime_rl/orchestrator/filters.py [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+                    consecutive = 0
+                if consecutive >= self.window:
+                    return FilterResult(detected=True, detection_index=global_idx)
+                global_idx += 1
+        return FilterResult(detected=False)
+
+
+def setup_filter(config: FilterConfigType, vocab_size: int) -> RolloutFilter:
+    """Create a RolloutFilter from a filter config."""
+    if isinstance(config, GibberishFilterConfig):
```
Comment (mikasenghaas) [inline src/prime_rl/orchestrator/filters.py]: can use `config.type == ...` here and get strong typing after

File: src/prime_rl/orchestrator/filters.py [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+        return {}
+
+    counts: dict[str, int] = {f.name: 0 for f in filters}
+    total_detected = 0
+    total_enforced = 0
+
+    for rollout in rollouts:
+        if rollout.get("metrics") is None:
+            rollout["metrics"] = {}
+        for filt in filters:
```
Comment (mikasenghaas) [inline src/prime_rl/orchestrator/filters.py:146]: lol any reason to not name it `filter` here?

File: src/prime_rl/orchestrator/filters.py [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+                            tokens["completion_mask"] = [0] * len(tokens["completion_mask"])
+                    rollout["stop_condition"] = filt.name
+                    total_enforced += 1
+
+                break
+
+    n = len(rollouts)
+    metrics: dict[str, float] = {}
+    for f in filters:
+        metrics[f"filter/{f.name}_count"] = float(counts[f.name])
```
Comment (mikasenghaas) [inline src/prime_rl/orchestrator/filters.py:169]: more in line with metric logging convention would be `filter/<name>/count` and `filter/<name>/mean` but idk have not been too happy with our consistency in logging for some time so could also fix later haha

File: src/prime_rl/orchestrator/filters.py [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+                rollout["metrics"][f"filter/{filt.name}"] = 1.0
+                counts[filt.name] += 1
+                total_filtered += 1
+                break
+
+    n = len(rollouts)
+    metrics: dict[str, float] = {}
+    for f in filters:
+        metrics[f"filter/{f.name}_count"] = float(counts[f.name])
+        metrics[f"filter/{f.name}_rate"] = counts[f.name] / n if n > 0 else 0.0
```
[the remark this answers is not in the document]
Comment (mikasenghaas) [reply src/prime_rl/orchestrator/filters.py:170]: ah yea should maybe cover this edge case and ensure that the filter config types are unique which should ensure unique names as well right?

**Review by mikasenghaas — COMMENTED** [review]
really nice!

- we should probably aim to turn online difficulty filtering (currently part of buffer) into smth like `ZeroAdvantageFilter`. can do part of later pr tho
- can we do a hendrycks/acereason math run to double check the defaults are sensible (with enforce on)
- i think we change the reward in-place when we `apply_filter`. should think clearly about what behavior we want, e.g. should a filtered rollout contribute nothing to baseline or 0 or its original? same for simple logging of e.g. training reward should we include or exclude filtered rollouts?

**Comment by mikasenghaas** [comment]
ah thinking about the zero adv filter again: ig its qualitatively a bit different bc the filter is based on a group not an individual rollout. i think we might still want to consider if we can consolidate both abstractions. i really like the filter one so if it were able to handle this one as well would be great

File: src/prime_rl/orchestrator/config.py (hunk 7 above)
Comment (hallerite) [reply src/prime_rl/orchestrator/config.py:585]: exactly

File: src/prime_rl/orchestrator/config.py (hunk 8 above)
Comment (hallerite) [reply src/prime_rl/orchestrator/config.py]: true, should make description more parsimonious

File: src/prime_rl/orchestrator/config.py (hunk 9 above)
Comment (hallerite) [reply src/prime_rl/orchestrator/config.py:594]: I just followed the defaults of the paper, but we also make it a percentile

File: src/prime_rl/orchestrator/config.py (hunk 10 above)
Comment (hallerite) [reply src/prime_rl/orchestrator/config.py]: you are right, this is an artifact from an earlier version that didn't derive `vocab_size` automatically

File: src/prime_rl/orchestrator/filters.py (hunk 14 above)
Comment (hallerite) [reply src/prime_rl/orchestrator/filters.py:169]: I think we should think about this more generally, but probably better to make a separate PR to consolidate

File: src/prime_rl/orchestrator/filters.py (hunk 13 above)
Comment (hallerite) [reply src/prime_rl/orchestrator/filters.py:146]: yeah, `filter` is a Python builtin, so it's convention not too shadow it (even if harmless here).

File: src/prime_rl/orchestrator/filters.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (hallerite) [reply src/prime_rl/orchestrator/filters.py:170]: yes

**Review by samsja — APPROVED** [review]
